### ISSUE TITLE
State management with Riverpod

### ISSUE THREAD

**Issue by SimonLab** [dwyl/flutter-todo-list-tutorial#27]
Now that a first version of stage management is implement (see https://github.com/dwyl/flutter-todo-list-tutorial/issues/22) I want to look at Riverpod and understand how it works and how it can improve the application, see https://pub.dev/packages/riverpod

> This project can be considered as a rewrite of provider to make improvements that would be otherwise impossible.

**Comment by SimonLab** [comment]
I've started to change the code to use Riverpod (see #28). Creating Riverpod providers for the todolist itself work however I haven't managed to create providers for the items (too manage the "completed" status).
I've looked at the following example: https://github.com/rrousselGit/river_pod/tree/master/examples/todos but I need to understand more Riverpod concepts to undersatnd fully the code.
I'm watching the following serie: https://www.youtube.com/playlist?list=PLjr4ufdmNA4IjEQ6xIv_sQdyBhm2kn1Qv and update the code as I learn more about how to use Riverpod

**Comment by SimonLab** [comment]
Providers: Objects that contains value that can be listen to. When the values change the providers will notify the listening widgets. Different providers exist to be able to listen to different type of values/objects

  - Provider: read only for simple value
  - StateProvider
  - StateNotifierProvider (to be used with StateNotifier values)
  - FutureProvider
  - StreamProvider
  - ProviderScope
  - ScopedProvider
  - ChangeNotifierProvider (to be used with ChangeNotifier value. StateProvider/StateNotifierProvider can be used instead of this provider)

ref: 
- https://riverpod.dev/docs/concepts/providers: A good brief on what are providers
- https://developermemos.com/posts/different-providers-riverpod: Short explanations of the different providers
- https://pub.dev/documentation//flutter_riverpod/latest/flutter_riverpod/flutter_riverpod-library.html : The flutter riverpod library documentation

**Comment by nelsonic** [comment]
@LuchoTurtle please research and confirm if `Riverpod` 
is still the "best" way to manage state in a `Flutter` app in 2022 
***or*** if the `Flutter` core now has this functionality 
***or*** there is another package that we could/should use instead. 🔍 

My reading of the https://pub.dev/packages/riverpod docs is that the project is well-maintained:
<img width="1314" alt="riverpod-web-mobile" src="https://user-images.githubusercontent.com/194400/200548013-c1d9e9f0-91bf-4ab2-b811-59644d317351.png">

but if there is something built-in that we can use instead, then we should consider it. 💭 

Thanks! 🙏

**Comment by LuchoTurtle** [comment]
I've talked with [a friend](https://github.com/freitzzz) about this, as he has worked with Flutter for much more time and he **highly** recommended [`Bloc`](https://bloclibrary.dev/#/) as a state management tool. 
In fact, it seems the community usually gravitate between `Riverpod` and `Bloc` quite often:
- https://www.reddit.com/r/FlutterDev/comments/pjj40p/riverpod_vs_bloc/
- https://www.reddit.com/r/FlutterDev/comments/kyd7zs/riverpod_a_better_state_management_solution_than/
- https://www.reddit.com/r/FlutterDev/comments/wmh2oh/eli5_bloc_compared_to_providerriverpod/

and both in fact are the *gold standard* when it comes to state management. 
However, some differences do exist. `Riverpod` appears to be a `Provider` 2.0 while Bloc is an opinionated framework. `Bloc` **forces** you to separate UI and logic while `Riverpod` is more flexible. There are also arguments that `Bloc` offers more value as the project grows and `Riverpod` allows for faster iteration on smaller projects. 

People also say Riverpod's documentation and testing are lacking, when compared with Bloc.

People say Bloc is too opinionated and rigid.

[It's an endless discussion](https://www.google.com/search?q=riverpod+vs+bloc&oq=riverpod+vs+bloc&aqs=chrome.0.69i59l3j0i20i263i512j0i22i30j69i60l2j69i61.2893j0j1&sourceid=chrome&ie=UTF-8)

We can compare the activity and momentum between Riverpod and Bloc by just looking at their repos
https://github.com/felangel/bloc
https://github.com/rrousselGit/riverpod

`Bloc` seems to take the cake.

Food for thought.

**Comment by nelsonic** [comment]
@LuchoTurtle happy to trust @freitzzz experience on this one; seems to know what he's doing. 🥋
Thank you for reaching out to him for expert insight. Good to see you leveraging your network. 👌
And thanks for sharing links to discussions on Reddit. 👀

**Comment by nelsonic** [comment]
https://github.com/dwyl/flutter-todo-list-tutorial/pull/28 merged. ✅
